activerecord/lib/active_record/relation.rb

```diff
@@ -642,6 +642,13 @@ def inspect
       "#<#{self.class.name} [#{entries.join(', ')}]>"
     end
 
+    def empty_scope? # :nodoc:
+      @values.empty? ||
```

Comment:
I think it's a premature optimization. I'd rather not be so tightly coupled to STI unless we absolutely have to